embark.el

```diff
@@ -1499,20 +1496,21 @@ the target at point."
                                   (last-nonmenu-event 13))
                               (setq prefix-arg prefix)
                               (embark--run-action-hooks embark-pre-action-hooks
-                                                        action target bounds)
+                                                        action target quit)
```

Comment:
That's the good thing there - we can add arguments in the future. However the allow-edit parameter seems unnecessary given action and the embark-allow-edit-command lists?

Reply:
It is strictly speaking unnecessary, but since the actual logic to figure out if editing is allowed is a little tedious (if editing is allowed by default check the blacklist, if not check the whitelist), it would be a convenience for users.

Reply:
Yes, but let's not add convenience until actually requested ;) The cleaner solution would be to write a function `embark--allow-edit-p` which can be called on the :action value to hide the logic.

I also thought about maybe simplifying the variables and logic of these allow/disallow edit lists. Maybe we can unify the variables somehow. Embark has relatively many configuration variables, if we can cut this down I am all for it. The `t` and `:always` keys of the hooks are a step in the right direction.

Reply:
My point is precisely that it was requested! I can't remember where though. It's probably here on GitHub, in some issue.

Reply:
As for unifying the variables, we could have a single `embark-allow-edit` plist with keys `:yes`, `:no`, and `:default` (or maybe `:allow`, `:disallow` and `t`).

Reply:
> My point is precisely that it was requested! I can't remember where though. It's probably here on GitHub, in some issue.

Of course then we can add this. However my experience shows that people come to a project, make some request use it for a while and then stop using the project. So I am hesitant to retroactively add features as long as I am not convinced that they are strictly needed. In this case for example I am unconvinced - I believe all properties which can be derived from `:action` or the `:target` arguments should not be passed directly. Instead providing a function `embark--allow-edit-p` is a more scalable solution. Maybe you have a different opinion about this - I am trying to find some balance :)

> As for unifying the variables, we could have a single embark-allow-edit plist with keys :yes, :no, and :default (or maybe :allow, :disallow and t).

I already considered this. The downside is that it is much less user friendly since you cannot quickly switch from allow->disallow by setting a single flag. I haven't found a good solution yet.

Reply:
Yeah, the `embark--allow-edit-p` function sounds better. I have this really old-fashioned aversion to recomputing things, but you're right that it's better in this case.

Reply:
> I already considered this. The downside is that it is much less user friendly since you cannot quickly switch from allow->disallow by setting a single flag. I haven't found a good solution yet.

What do you mean exactly? In the proposal I'm making the value of the `:default` key plays the same role as the current `embark-allow-edit-default`.

Reply:
Oh sorry, this makes a lot of sense what you write. It is probably a good solution with the :default value. This is not something I considered. Btw, do you think it makes sense to use keywords here? I think keywords are a better fit for keys and they should mainly be used in plists (not as alist keys) and of course for cl-defun keys.

Reply:
Oh, I think I totally misunderstood your proposal. I don't think using a plist is any easier than multiple variables.

Reply:
What about this?

~~~ elisp
(defun embark-edit-action-alist
  '((command . nil)         ;; no edit
    (command . t)           ;; allow edit
    (command . default)     ;; use default value
    (t . nil))              ;; set to t=enable, nil=disable by default
 ~~~

Using `t` as key is a bit weird here but consistent with how we use `t` throughout Embark. Alternatively:


~~~ elisp
(defun embark-edit-action-alist
  '((command . nil)         ;; no edit
    (command . t)           ;; allow edit
    (command . default)     ;; use default value
    (default . nil))        ;; set to t=enable, nil=disable by default
 ~~~

Reply:
You're right: a plist grouping several variables is not easier than several variables, just maybe a little tidier. You seemed to like the same proposal when I made it for embark-quit-after-action. :)

A real simplification would involve simplifying the logic too.

Reply:
> You're right: a plist grouping several variables is not easier than several variables, just maybe a little tidier. You seemed to like the same proposal when I made it for embark-quit-after-action. :)

What did I like exactly? Maybe I am a bit inconsistent or my opinion changed. I am not that of an experienced Elisp coder. I never wrote any serious Elisp before Consult. :-P

Reply:
Well, experienced or not I defer to you in all matters of Emacs Lisp coding style. Well, except the width of lines and how often one should use `cl-loop`... 😄

Reply:
Also about spaces in docstrings after dots? Preparing a PR.... :laughing:

Reply:
I only uses double spaces in docstring because I like some of the other things checkdoc checks for...

Reply:
I also use checkdoc/package-lint. checkdoc respects `sentence-end-double-space` :-P

Reply:
I don't think it does, otherwise why does it keep pestering me about the docstrings? But even if it does, I want my documentation to look like the Emacs documentation. :P

Reply:
Hmm, didn't you say that you set `sentence-end-double-space` differently in elisp files? But I agree with the consistency throughout Emacs.

Reply:
Also, Emacs is a weird system. It only fits to use a weird double space convention.